spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/CloudSqlEnvironmentPostProcessor.java

```diff
@@ -64,55 +64,116 @@ public void postProcessEnvironment(ConfigurableEnvironment environment, SpringAp
+		// Configure default JDBC driver and username as fallback values when not specified
+		Map<String, Object> fallbackMap = new HashMap<>();
+		fallbackMap.put("spring.datasource.username", databaseType.getDefaultUsername());
+		fallbackMap.put("spring.datasource.driver-class-name", cloudSqlJdbcInfoProvider.getJdbcDriverClass());
+		environment.getPropertySources()
+				.addLast(new MapPropertySource("CLOUD_SQL_DATA_SOURCE_FALLBACK", fallbackMap));
 
-			// always set the spring.datasource.url property in the environment
-			Map<String, Object> primaryMap = new HashMap<>();
-			primaryMap.put("spring.datasource.url", cloudSqlJdbcInfoProvider.getJdbcUrl());
-			environment.getPropertySources()
-					.addFirst(new MapPropertySource("CLOUD_SQL_DATA_SOURCE_URL", primaryMap));
+		// Always set the spring.datasource.url property in the environment
+		Map<String, Object> primaryMap = new HashMap<>();
+		primaryMap.put("spring.datasource.url", cloudSqlJdbcInfoProvider.getJdbcUrl());
+		environment.getPropertySources()
+				.addFirst(new MapPropertySource("CLOUD_SQL_DATA_SOURCE_URL", primaryMap));
+	}
 
-			setCredentials(sqlProperties, gcpProperties);
 
-			// support usage metrics
-			CoreSocketFactory.setApplicationName("spring-cloud-gcp-sql/"
-					+ this.getClass().getPackage().getImplementationVersion());
+	/**
+	 * Sets the R2DBC properties such as username and connection url based on the Cloud SQL
+	 * properties specified.
+	 * @param environment environment to post-process
+	 * @param sqlProperties cloud sql properties
+	 * @param r2dbcDatabaseType enum containing MySQl and PostgreSQL information (for example,
+	 *     r2dbc url and default username)
+	 */
+	private void applyR2dbcSettings(ConfigurableEnvironment environment, GcpCloudSqlProperties sqlProperties,
+									R2dbcDatabaseType r2dbcDatabaseType) {
+		String r2dbcUrl = createR2dbcUrl(r2dbcDatabaseType, sqlProperties);
+		if (LOGGER.isInfoEnabled()) {
+			LOGGER.info("Default " + r2dbcDatabaseType.name()
+					+ " R2dbcUrl provider. Connecting to "
+					+ r2dbcUrl);
 		}
+
+		// Add default username as fallback when not specified
+		Map<String, Object> fallbackMap = new HashMap<>();
+		fallbackMap.put("spring.r2dbc.username", r2dbcDatabaseType.getDefaultUsername());
+		environment.getPropertySources().addLast(new MapPropertySource("CLOUD_SQL_R2DBC_FALLBACK", fallbackMap));
+
+		// Always set the spring.r2dbc.url property in the environment
+		Map<String, Object> primaryMap = new HashMap<>();
+		primaryMap.put("spring.r2dbc.url", r2dbcUrl);
+		environment.getPropertySources().addFirst(new MapPropertySource("CLOUD_SQL_R2DBC_URL", primaryMap));
 	}
 
-	private DatabaseType getEnabledDatabaseType(ConfigurableEnvironment environment) {
+	String createR2dbcUrl(R2dbcDatabaseType databaseType, GcpCloudSqlProperties sqlProperties) {
+		Assert.hasText(sqlProperties.getDatabaseName(), "A database name must be provided.");
```

Comment:
Currently holding off on adding support for enable IAM auth through the url as this doesn't seem to be working as expected for the R2DBC workflow. Extra options specified through the url are automatically typed as `String` so  when [GcpConnectionFactoryProvider#createFactory](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/20881af5ec8669fcb97899f232810ed03138ebfa/r2dbc/core/src/main/java/com/google/cloud/sql/core/GcpConnectionFactoryProvider.java#L103) calls `(Boolean)connectionFactoryOptions.getValue(ENABLE_IAM_AUTH)` this results in an error.

Reply:
Created https://github.com/GoogleCloudPlatform/spring-cloud-gcp/issues/821 to track this.